File: backend/soul_reload.ts
/**
 * soul_reload.ts
 * Hot-reload handler for SOUL.md updates received via WebSocket.
 * Called when event == "agent_soul_updated".
 *
 * This module is designed for OpenClaw gateway integration.
 * Copy this file to: openclaw/src/gateway/soul_reload.ts
 */
import { WebSocket } from 'ws';
import { logger } from '../utils/logger';

// Default ClawController API host
const DEFAULT_CLAW_CONTROLLER_HOST = 'localhost:8000';

export interface AgentSoulUpdatedEvent {
  event: 'agent_soul_updated';
  agent_id: string | number;
  score?: number;
}

export interface SessionManager {
  getSession(agentId: string | number): any | null;
}

/**
 * Handle hot-reload of agent SOUL.md when OpenClaw receives 'agent_soul_updated' event.
 *
 * This function:
 * 1. Extracts agent_id from the WebSocket event
 * 2. Finds the active session for that agent
 * 3. Fetches updated SOUL.md from ClawController API
 * 4. Applies the new SOUL to the running session without restarting
 *
 * @param event - The full WebSocket message dict
 * @param sessionManager - The OpenClaw session manager object
 * @returns Promise<void>
 *
 * @example
 * // Event shape:
 * {
 *   "event": "agent_soul_updated",
 *   "agent_id": "dev",
 *   "score": 87
 * }
 */
export async function handle_soul_updated(
  event: AgentSoulUpdatedEvent,
  sessionManager: SessionManager
): Promise<void> {
  const { agent_id, score } = event;

  // Extract agent_id from event
  if (!agent_id) {
    logger.warn('SOUL reload: Missing agent_id in agent_soul_updated event');
    return;
  }

  // Get active session for this agent
  const session = sessionManager.getSession(agent_id);

  if (!session) {
    logger.info(
      `No active session for agent ${agent_id}, skipping SOUL reload (score: ${score ?? 'N/A'})`
    );
    return;
  }

  // Get ClawController host from environment or use default
  const clawHost = getClawControllerHost();

  // Fetch updated SOUL from ClawController API
  try {
    const url = `http://${clawHost}/api/agents/${agent_id}/soul`;
    const response = await fetch(url, {
      method: 'GET',
      headers: {
        'Content-Type': 'application/json',
      },
    });

    if (!response.ok) {
      logger.error(
        `SOUL reload failed for agent ${agent_id}: HTTP ${response.status} from ClawController`
      );
      return;
    }

    const data = await response.json();
    const soulMarkdown = data.soul_markdown || data.soul || '';

    if (!soulMarkdown) {
      logger.error(
        `SOUL reload failed for agent ${agent_id}: No soul_markdown field in response`
      );
      return;
    }

    // Apply new SOUL to running session
    if (typeof session.reloadSoul === 'function') {
      session.reloadSoul(soulMarkdown);
    } else {
      logger.error(
        `SOUL reload failed for agent ${agent_id}: Session does not have reloadSoul method`
      );
      return;
    }

    logger.info(`✅ SOUL hot-reloaded for agent ${agent_id} (score: ${score ?? 'N/A'})`);
  } catch (error) {
    if (error instanceof Error) {
      if (error.message.includes('fetch') || error.message.includes('ECONNREFUSED')) {
        logger.error(
          `SOUL reload failed for agent ${agent_id}: Cannot connect to ClawController at ${clawHost} - ${error.message}`
        );
      } else if (error.message.includes('timeout')) {
        logger.error(
          `SOUL reload timeout for agent ${agent_id}: ClawController at ${clawHost} did not respond - ${error.message}`
        );
      } else {
        logger.error(
          `SOUL reload error for agent ${agent_id}: ${error.name} - ${error.message}`
        );
      }
    } else {
      logger.error(
        `SOUL reload error for agent ${agent_id}: Unknown error - ${String(error)}`
      );
    }
  }
}

/**
 * Get ClawController API host from environment variable.
 *
 * @returns Host string (e.g., "localhost:8000")
 *
 * @env CLAW_CONTROLLER_HOST - Override default host (default: "localhost:8000")
 */
function getClawControllerHost(): string {
  return process.env.CLAW_CONTROLLER_HOST || DEFAULT_CLAW_CONTROLLER_HOST;
}

File: backup_soul_reload_20260301_045325/openclaw/soul_reload.py
"""
soul_reload.py
Hot-reload handler for SOUL.md updates received via WebSocket.
Called when event == "agent_soul_updated".

This module is designed for OpenClaw gateway integration.
Copy this file to: openclaw/soul_reload.py
"""
import logging
from typing import Any, Dict, Optional

import httpx

logger = logging.getLogger(__name__)

# Default ClawController API host
DEFAULT_CLAW_CONTROLLER_HOST = "localhost:8000"


async def handle_soul_updated(event: Dict[str, Any], session_manager: Any) -> None:
    """
    Handle hot-reload of agent SOUL.md when OpenClaw receives 'agent_soul_updated' event.

    This function:
    1. Extracts agent_id from the WebSocket event
    2. Finds the active session for that agent
    3. Fetches updated SOUL.md from ClawController API
    4. Applies the new SOUL to the running session without restarting

    Args:
        event: The full WebSocket message dict containing:
            - event: "agent_soul_updated"
            - agent_id: int (the agent identifier)
            - score: int (optional, for logging)
        session_manager: The OpenClaw session manager object (passed in, not imported)

    Returns:
        None

    Example event:
        {
            "event": "agent_soul_updated",
            "agent_id": 42,
            "score": 87
        }
    """
    # Extract agent_id from event
    agent_id = event.get("agent_id")
    score = event.get("score", "N/A")

    if agent_id is None:
        logger.warning("SOUL reload: Missing agent_id in agent_soul_updated event")
        return

    # Get active session for this agent
    session = session_manager.get_session(agent_id)

    if session is None:
        logger.info(
            f"No active session for agent {agent_id}, skipping SOUL reload "
            f"(score: {score})"
        )
        return

    # Get ClawController host from environment or use default
    claw_host = _get_claw_controller_host()

    # Fetch updated SOUL from ClawController API
    async with httpx.AsyncClient(timeout=10.0) as client:
        try:
            url = f"http://{claw_host}/api/agents/{agent_id}/soul"
            response = await client.get(url)

            if response.status_code != 200:
                logger.error(
                    f"SOUL reload failed for agent {agent_id}: "
                    f"HTTP {response.status_code} from ClawController"
                )
                return

            data = response.json()
            soul_markdown = data.get("soul_markdown") or data.get("soul", "")

            if not soul_markdown:
                logger.error(
                    f"SOUL reload failed for agent {agent_id}: "
                    f"No soul_markdown field in response"
                )
                return

            # Apply new SOUL to running session
            session.reload_soul(soul_markdown)

            logger.info(
                f"✅ SOUL hot-reloaded for agent {agent_id} (score: {score})"
            )

        except httpx.ConnectError as e:
            logger.error(
                f"SOUL reload failed for agent {agent_id}: "
                f"Cannot connect to ClawController at {claw_host} - {e}"
            )
        except httpx.TimeoutException as e:
            logger.error(
                f"SOUL reload timeout for agent {agent_id}: "
                f"ClawController at {claw_host} did not respond - {e}"
            )
        except Exception as e:
            logger.error(
                f"SOUL reload error for agent {agent_id}: {type(e).__name__} - {e}"
            )


def _get_claw_controller_host() -> str:
    """
    Get ClawController API host from environment variable.

    Returns:
        Host string (e.g., "localhost:8000" or "api.clawcontroller.com")

    Environment:
        CLAW_CONTROLLER_HOST: Override default host (default: "localhost:8000")
    """
    import os
    return os.environ.get("CLAW_CONTROLLER_HOST", DEFAULT_CLAW_CONTROLLER_HOST)
